# ==============================================================================
# FIXING COMMIT DIFF
# ==============================================================================
# File: docshell/base/BrowsingContextGroup.cpp
# Commit: fa7bf737d4eb
# Full Hash: fa7bf737d4ebfe0394abf08083d312c750076952
# Author: Nika Layzell <nika@thelayzells.com>
# Date: 2020-11-27 15:53:21
# Description:
#   Bug 1676892 - Remove non-subscribed hosts when destroying BrowsingContextGroup, r=farre
#   
#   Before this change, a BrowsingContextGroup would fail to remove itself from a
#   ContentParent's mGroups table if the process was still launching when the BCG
#   was destroyed. This was because until the ContentParent had launched, it existed
# ==============================================================================

diff -r 3636cdf0b487 -r fa7bf737d4eb docshell/base/BrowsingContextGroup.cpp
--- a/docshell/base/BrowsingContextGroup.cpp	Thu Nov 26 16:19:35 2020 +0000
+++ b/docshell/base/BrowsingContextGroup.cpp	Thu Nov 26 06:58:29 2020 +0000
@@ -228,10 +228,16 @@
   mDestroyed = true;
 #endif
 
-  mHosts.Clear();
+  // Make sure to call `RemoveBrowsingContextGroup` for every entry in both
+  // `mHosts` and `mSubscribers`. This will visit most entries twice, but
+  // `RemoveBrowsingContextGroup` is safe to call multiple times.
+  for (auto& entry : mHosts) {
+    entry.GetData()->RemoveBrowsingContextGroup(this);
+  }
   for (auto& entry : mSubscribers) {
     entry.GetKey()->RemoveBrowsingContextGroup(this);
   }
+  mHosts.Clear();
   mSubscribers.Clear();
 
   if (sBrowsingContextGroups) {
